include/aca_arp_responder.h

```diff
@@ -0,0 +1,148 @@
+// Copyright 2019 The Alcor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ACA_ARP_RESPONDER_H
+#define ACA_ARP_RESPONDER_H
+
+#include <string>
+#include <unordered_map>
+#include "hashmap/HashMap.h"
+#include <mutex>
+
+using namespace std;
+
+namespace aca_arp_responder
+{
+
+#define ARP_ENTRY_DATA_SET(pData,pConfig)                           \
+  do {                                                              \
+    (pData)->ipv4_address = (pConfig)->ipv4_address;                \
+    (pData)->vlan_id = (pConfig)->vlan_id;                          \
+  } while (0)                                                       
+
+#define ARP_TABLE_DATA_SET(pData,pConfig)                           \
+  do {                                                              \
+    (pData)->mac_address = (pConfig)->mac_address;                  \
+  } while (0)  
+
+
+struct arp_config{
+  string mac_address;
+  string ipv4_address;
+  string ipv6_address; 
+  uint16_t vlan_id;
+  string port_host_name;
+};
+
+struct arp_entry_data{
+  string ipv4_address;
+  string ipv6_address; 
+  uint16_t vlan_id;
+  bool operator==(const arp_entry_data &p) const{
+    return ipv4_address==p.ipv4_address && vlan_id==p.vlan_id;
+  }
+
+  bool operator!=(const arp_entry_data &p) const{
+    return ipv4_address!=p.ipv4_address || vlan_id!=p.vlan_id;
+  }
+};
+
+
+struct arp_hash{
+  size_t operator() (const arp_entry_data &p) const{
+    return hash<string>()(p.ipv4_address) ^ (hash<uint16_t>()(p.vlan_id) << 1);
+  }
+};
+
+struct arp_table_data{
+  string mac_address;
+  mutable std::shared_timed_mutex arp_mutex;
```

Comment:
not sure if we want to track a mutex in each arp_table_data. I suggest switching to use our new concurrent hashtable which does the locks for you, see below in aca_vlan_manager.h for reference:

  // hashtable <key: tunnel ID, value: vpc_table_entry>
  CTSL::HashMap<uint, vpc_table_entry *> _vpcs_table;

Reply:
Ok, I see. It seems that we don't need mutex while operating different entries, but is there any chance that multiple threads will write simultaneously in the same entry?